### ISSUE TITLE
clang mis-compilation with SFINAE/enable_if

### ISSUE THREAD

**Issue by njames93** [llvm/llvm-project#62466]
We recently tried to add clang as a supported compiler for a project, however we noticed what appears to be a nasty bug with overload resolution. Take this reduced example here
```cpp
#include <type_traits>

struct B {
    template <typename T>
    operator T() const;
};

struct C {
    template <typename T>
    operator T() const;
};

template <typename Base = C>
struct A : Base {
    template <typename T, typename B = Base,
              typename = std::enable_if_t<std::is_same_v<B, C>>>
    operator T() const;
};

enum class Enum {};

void foo() {
    A<B> a;
    Enum b;
    b = a;  // Error cause
}
```
In this example, during overload resolution, The `operator T()` in `struct A` has a substitution failure due to the failed `enable_if` condition, therefore it is disregarded as a candidate. However, in its base struct,`B`, there is a valid candidate with no restrictions that should be selected.
Both MSVC and GCC latest release and trunk compile this code correctly, clang 16.0 and trunk both reject this
https://godbolt.org/z/b1nf68bds

**Comment by shafik** [comment]
Yeah, this looks like a clang bug, I tried replacing it w/ a `requires` but no luck there either.

CC @erichkeane to see if this is covered by another bug

**Comment by erichkeane** [comment]
I don't have a good idea here, though I would not be surprised if there is a CWG issue about the operator being hidden?  So it isn't clear to me that this is valid code (or perhaps just us not having a DR implemented/vice versa).

I'll note, EDG ALSO rejects this for the same reason as clang: https://godbolt.org/z/zsb8cE5Ks

**Comment by zygoloid** [comment]
This looks related to #58571 -- I think Clang is incorrectly determining that the derived `operator T` hides the base `operator T` despite them having different *template-head*s. I think the rule in this case changed in [P1787R6](http://wg21.link/p1787r6); the rules now say:

> A conversion function template with a dependent return type hides only templates in base classes that correspond to it ([class.member.lookup])

Do we have any tracking of our implementation status for P1787R6?

**Comment by shafik** [comment]
@Endilll has been working through all the DRs in P1787R6 and updating their statues and adding tests. I think that is all we have at this point.

**Comment by Endilll** [comment]
Yes, I've been working on tests for P1787R6, but I haven't touched template-related CWG issues mentioned there, which I guess this issue falls under.
